Fix issue 3491

PalAditya opened the pull request:
![Error](https://user-images.githubusercontent.com/25523604/64914093-c8170300-d769-11e9-8321-fb64e19fcee8.PNG)

![Solution](https://user-images.githubusercontent.com/25523604/64914094-c8170300-d769-11e9-9980-0c52d87d976b.PNG)

![Capture](https://user-images.githubusercontent.com/25523604/64914337-e92d2300-d76c-11e9-9193-59216a296385.PNG)

Solves issue #3491 . Please find a sample execution before and after the changes

**Note** : The recommended solution in the thread (using *try-except* blocks) failed in **windows 10 1903, python 3.7.4** and it seems to be an issue with networkx itself as other `AttributeError` conditions are caught by python just fine. Hence, I went with the method of detecting versions, which solves it without any such problems. 😄 

Do let me know if this idea is intractable for some reason.

dschult commented:
Thanks for this!!

It's very helpful to use copy/paste of text rather than screen shots in your postings because we can't copy/paste the code from your screen shots.

What happens with the try-except code?

PalAditya commented:
If I use:
  ```
  try:
        lines = (line.decode(encoding) for line in path)
 except AttributeError:
        lines = (line for line in path)
```

Here's the stacktrace:
```
File "<stdin>", line 2, in <module>
File "<C:\Users\Lenovo\AppData\Local\Programs\Python\Python37\lib\site-packages\decorator.py:decorator-gen-626>", line 2, in read_edgelist
File "E:\GitHub\networkx\networkx\utils\decorators.py", line 240, in _open_file
result = func_to_be_decorated(*new_args, **kwargs)
File "E:\GitHub\networkx\networkx\readwrite\edgelist.py", line 382, in read_edgelist
data=data)
File "E:\GitHub\networkx\networkx\readwrite\edgelist.py", line 250, in parse_edgelist
for line in lines:
File "E:\GitHub\networkx\networkx\readwrite\edgelist.py", line 371, in <genexpr>
    lines = (line.decode(encoding) for line in path)
AttributeError: 'str' object has no attribute 'decode'
```
So, it **can** detect that it **is** an attribute error, but fails to catch it. I am unsure why; it can catch an equivalent `AttributeError` if I use str.decode() on a simple string.

dschult commented:
If it detects the AttributeError and it is inside a try/except clause it will catch the error (unless you spell the except clause incorrectly...:) So, there is something happening that you don't think is happening. 

In this case, the iterator isn't actually traversed until after you leave the try/except clause.
So the AttributeError occurs after you are into the parse_edgelist function.  You can make
the tuple-comprehension into a list-comprehension by using square brackets -- though for large
files that might use a lot of memory. 

So... let's try another approach:  Use an "if" clause inside the tuple-comprehension. Does this even work?

    lines = (line if isinstance(line, str) else line.decode(encoding) for line in path)

PalAditya commented:
Okay, that issue of iterator makes sense 😅 
Also, thanks for the brilliant idea of tuple comprehension: tested on my local machine, and it works brilliantly.
Updated the PR with the tuple comprehension method and removed redundant imports.

dschult commented:
Great!  Now for the tests -- :)
The tests are in tests/test_edgelist.py
Can you create one that tests reading in a file using mode 'r' instead of 'rb'?
The graph can be simple. The key is to make sure the decode doesn't raise an exception.

PalAditya commented:
Added the test case test_edge `test_read_edgelist_4()` which simulates reading a file in both **r** and **rb** mode, by using the **StringIO** and **BytesIO** methods respectively. I believe that this is what you wanted, and not read in directly from a file, right? 😅 

Although, even that can be added with something like:

```python
with open("a.txt","w") as f:
    f.write("1 0")
with open("a.txt","r") as f:
    G = nx.read_edgelist(f, nodetype=int, data=False)
assert_edges_equal(G.edges(), [(1 0)])
with open("a.txt","rb") as f:
    G = nx.read_edgelist(f, nodetype=int, data=False)
assert_edges_equal(G.edges(), [(1 0)])
os.remove("a.txt")
```
Please let me know what you think

dschult commented:
This is the approach I was intending. Thanks for all of this!
